auth/user_mgt.go

```diff
@@ -0,0 +1,553 @@
+		delete(p.payload, payloadName)
+	}
+	return nil
+}
+
+func (p *commonParams) addToListParam(listname, param string) {
+	if _, ok := p.payload[listname]; ok {
+		p.payload[listname] = append(p.payload[listname].([]string), param)
+	} else {
+		p.set(listname, []string{param})
+	}
+}
+
+func processClaims(p *commonParams, payloadName string) error {
+	if _, ok := p.payload["customClaims"]; !ok {
+		return nil
+	}
+	cc := p.payload["customClaims"]
+	claims, ok := cc.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("CustomClaims(): unexpected type")
+	}
+	for _, key := range reservedClaims {
+		if _, ok := claims[key]; ok {
+			return fmt.Errorf("CustomClaims(%q: ...): claim %q is reserved, and must not be set", key, key)
+		}
+	}
+	b, err := json.Marshal(claims)
+	if err != nil {
+		return fmt.Errorf("CustomClaims() Marshaling error: %v", err)
+	}
+	s := string(b)
+	if s == "null" {
+		s = "{}"
+	}
+	p.payload["customAttributes"] = s
+	delete(p.payload, "customClaims")
+	return nil
+}
+
+// Validators.
+
+// No validation needed. Used for bool fields.
+func validateTrue(p *commonParams, _ string) error {
+	return nil
+}
+
+func validateNonEmpty(p *commonParams, fieldName string) error {
+	if val, ok := p.payload[fieldName]; ok {
+		if len(val.(string)) == 0 {
+			return fmt.Errorf("%s must be a non-empty string", fieldName)
+		}
+	}
+	return nil
+}
+
+func validateEmail(p *commonParams, fieldName string) error {
+	if val, ok := p.payload[fieldName]; ok {
+		if parts := strings.Split(val.(string), "@"); len(parts) != 2 || len(parts[0]) == 0 || len(parts[1]) == 0 {
+			return fmt.Errorf("malformed Email string: %q", val)
+		}
+	}
+	return nil
+}
+
+func validatePassword(p *commonParams, fieldName string) error {
+	wantLength := 6
+	if val, ok := p.payload[fieldName]; ok {
+		if len(val.(string)) < wantLength {
+			return fmt.Errorf("Password must be a string at least %d characters long", wantLength)
+		}
+	}
+	return nil
+}
+
+func validateUID(p *commonParams, fieldName string) error {
+	wantLength := 128
+	if val, ok := p.payload[fieldName]; ok {
+		if len(val.(string)) > wantLength {
+			return fmt.Errorf("localId must be a string at most %d characters long", wantLength)
+		}
+	}
+	return nil
+}
+
+func validateCustomAttributes(p *commonParams, fieldName string) error {
+	wantLength := maxLenPayloadCC
+	if val, ok := p.payload[fieldName]; ok {
+		if len(val.(string)) > wantLength {
+			return fmt.Errorf("stringified JSON of CustomClaims must be a string at most %d characters long", wantLength)
+		}
+	}
+	return nil
+}
+
+func validatePhone(p *commonParams, fieldName string) error {
+	if val, ok := p.payload[fieldName]; ok {
+		if !regexp.MustCompile(`\+.*[0-9A-Za-z]`).MatchString(val.(string)) {
+			return fmt.Errorf(
+				"invalid PhoneNumber %q. Must be a valid, E.164 compliant identifier", val)
+		}
+	}
+	return nil
+}
+
+func (p *UserToCreate) preparePayload() (map[string]interface{}, error) {
+	if p.payload == nil {
```

Comment:
Can we make this (and also the other preparePayload), side-effect free? Currently these functions mutate the underlying `payload` map. Can we instead create a new map, and return it? Otherwise it's hard to reason about what happens if the developer uses the same `UserToCreate` or `UserToUpdate` instance in multiple calls.